src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -338,11 +338,23 @@ internal NamespaceOrTypeOrAliasSymbolWithAnnotations BindNamespaceOrTypeOrAliasS
             {
                 case SyntaxKind.NullableType:
                     {
-                        TypeSyntax typeArgumentSyntax = ((NullableTypeSyntax)syntax).ElementType;
+                        var nullableSyntax = (NullableTypeSyntax)syntax;
+                        TypeSyntax typeArgumentSyntax = nullableSyntax.ElementType;
                         TypeSymbolWithAnnotations typeArgument = BindType(typeArgumentSyntax, diagnostics, basesBeingResolved);
                         TypeSymbolWithAnnotations constructedType;
                         if (Compilation.IsFeatureEnabled(MessageID.IDS_FeatureStaticNullChecking))
                         {
+                            if (InExecutableBinder)
+                            {
+                                // Inside a method body or other executable code, we can afford to pull on NonNullTypes symbol or question IsReferenceType without causing cycles.
+                                // Types created outside executable context should be check by the responsible symbol (the method symbol checks its return type, for instance).
+
+                                if (typeArgument.IsReferenceType && NonNullTypesContext.NonNullTypes != true)
```

Comment:
Should this be `!typeArgument.IsValueType` to report warning on `T?` for unconstrained `T`?

Reply:
Makes sense. Thanks

---
In reply to: [205303506](https://github.com/dotnet/roslyn/pull/28842#discussion_r205303506) [](ancestors = 205303506)